Shopify/vscode-ruby-lsp
# Implement internal Ruby activation mechanism

### Motivation

Closes Shopify/ruby-lsp#1776, closes Shopify/ruby-lsp#1784, closes Shopify/ruby-lsp#1794, closes https://github.com/Shopify/ruby-lsp/issues/1216, closes Shopify/ruby-lsp#1762

Currently, we activate the Ruby environment using the user's version manager invoked through a shell. This is necessary because most version managers are shell scripts loaded in config files like `~/.zshrc` or `~/.bashrc`.

Unfortunately, this has been causing headache because some shell plugins or configurations may accidentally block or interact with our activation script, which leads to difficult to debug issues.

I believe we can provide a better experience if we provide a minimal Ruby environment activation mechanism implemented directly inside TypeScript so that we can avoid interacting with the shell altogether.

### Possible drawbacks

While this will fix the issues related to interacting with the shell, it's not free of drawbacks. Certain version managers provide more complex functionality that may not be supported by our internal mechanism.

For example, `rtx` supports adding extra environment variables in its `.rtx.toml` file. Other version managers allow users to change the default `GEM_HOME` where user gems are installed.

Moving forward with this approach means that we might need to add some extra complexity to handle this scenarios or at least make certain things configurable.

### Release strategy

We will need to have a long preview release cycle for this one to ensure that activation is working fine for common setups.

### Implementation

The bulk of the changes are in `Ruby`. All other changes are just putting pieces together.

The idea of the activation mechanism is that we

1. Try to determine the Ruby version by looking into common config files (e.g.: `.ruby-version`, `dev.yml`, `.tools-versions`, `.rtx.toml`)
2. We then try to find a Ruby installation for that version in the directories common used by version managers
3. If we can't find a Ruby version or can't find the directory for the installation, we prompt users to manually select their Ruby binary
4. We then run a script that basically just returns `GEM_HOME` and `GEM_ROOT` to us
5. Finally, we modify the environment to include these
    - GEM_HOME -> must point to the user directory for gem installations
    - GEM_ROOT -> must point to the directory for default gem installations
    - GEM_PATH -> a concatenation of GEM_HOME and GEM_ROOT
    - RUBY_VERSION - the Ruby version
    - PATH -> must contain the executable paths for user gems, default gems and for the Ruby `bin` folder

### Automated Tests

Added tests for all different config files we support.

### Manual Tests

We need to ensure that Ruby environment activation works properly for `chruby`, `rbenv`, `rvm`, `shadowenv`, `rtx` and `asdf`.

For all of them, the manual tests are analogous:

1. Write the configuration file with a Ruby version (e.g.: `.ruby-version`)
2. Start the Ruby LSP on that directory
3. Verify that the right Ruby version was picked
4. Verify that the LSP boots properly
